cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -346,6 +367,140 @@ size_type aggregate_reader_metadata::calc_num_row_groups() const
+      // translate chunk-relative row nums to absolute within the file
+      auto const pg_start_row = chunk_start_row + page_loc.first_row_index;
+      auto const pg_end_row =
+        chunk_start_row + (pg_idx == (num_pages - 1)
+                             ? rg.num_rows
+                             : offset_index.page_locations[pg_idx + 1].first_row_index);
+
+      auto const num_rows = pg_end_row - pg_start_row;
+      page_info pg_info{page_loc, num_rows};
+
+      // check to see if we already have null counts for each page
+      if (column_index.null_counts.has_value()) {
+        pg_info.num_nulls = column_index.null_counts.value()[pg_idx];
+      }
+
+      // save variable length byte info if present
+      if (offset_index.unencoded_byte_array_data_bytes.has_value()) {
+        pg_info.var_bytes_size = offset_index.unencoded_byte_array_data_bytes.value()[pg_idx];
+      }
+
+      // if def histogram is present, then use it to calculate num_valid and num_nulls
+      if (def_hist != nullptr) {
+        auto const h      = &def_hist[pg_idx * (max_def_level + 1)];
+        pg_info.num_valid = h[max_def_level];
+
+        // calculate num_nulls if not available from column index
+        if (not pg_info.num_nulls.has_value()) {
+          pg_info.num_nulls = std::reduce(h, h + max_def_level);
+        }
+      }
+      // there is no def histogram.
+      // if there is no repetition (no lists), then num_values == num_rows, and num_nulls can be
+      // obtained from the column index
+      else if (max_rep_level == 0) {
+        // if we already have num_nulls from column index
+        if (pg_info.num_nulls.has_value()) {
+          pg_info.num_valid = pg_info.num_rows - pg_info.num_nulls.value();
+        }
+        // if max_def is 0, there are no nulls
+        else if (max_def_level == 0) {
+          pg_info.num_nulls = 0;
+          pg_info.num_valid = pg_info.num_rows;
+        }
+      }
+      // if the rep level histogram is present, we can get the total number of values
+      // from that
+      else if (rep_hist != nullptr) {
+        if (pg_info.num_nulls.has_value()) {
+          auto const h          = &rep_hist[pg_idx * (max_rep_level + 1)];
+          auto const num_values = std::reduce(h, h + max_rep_level + 1);
+          pg_info.num_valid     = num_values - pg_info.num_nulls.value();
+        }
+      }
+
+      // If none of the ifs above triggered, then we have neither histogram (likely the writer
+      // doesn't produce them, the r:0 d:1 case should have been handled above). The column index
+      // doesn't give us value counts, so we'll have to rely on the page headers. If the histogram
+      // info is missing or insufficient, then just return without modifying the row_group_info.
+      if (not pg_info.num_nulls.has_value() or not pg_info.num_valid.has_value()) { return; }
+
+      // if using older page indexes that lack size info, don't use
```

Comment:
```suggestion
      // if using older page indexes that lack size info, don't use `pg_info`
```

Reply:
Cleaned this up and added a more detailed TODO...will push once CI finishes

Reply:
looking at a few other PRs, this might be an "_if_ CI finishes" situation :grimacing: